std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
+const Token = std.zig.Token;
+const TokenIterator = Tree.TokenList.Iterator;
 
-/// Result should be freed with tree.deinit() when there are
-/// no more references to any of the tokens or nodes.
-pub fn parse(allocator: *mem.Allocator, source: []const u8) !*ast.Tree {
+pub const Error = error{UnexpectedToken} || Allocator.Error;
+
+pub fn parse(allocator: *Allocator, source: []const u8) !*Tree {
     var tree_arena = std.heap.ArenaAllocator.init(allocator);
     errdefer tree_arena.deinit();
+    const arena = &tree_arena.allocator;
 
-    var stack = std.ArrayList(State).init(allocator);
-    defer stack.deinit();
+    var token_list = Tree.TokenList.init(arena);
+    var tokenizer = std.zig.Tokenizer.init(source);
+    while (true) {
+        const tree_token = try token_list.addOne();
+        tree_token.* = tokenizer.next();
+        if (tree_token.id == .Eof) break;
+    }
+    var it = token_list.iterator(0);
 
-    const arena = &tree_arena.allocator;
-    const root_node = try arena.create(ast.Node.Root);
-    root_node.* = ast.Node.Root{
-        .base = ast.Node{ .id = ast.Node.Id.Root },
-        .decls = ast.Node.Root.DeclList.init(arena),
-        .doc_comments = null,
-        // initialized when we get the eof token
-        .eof_token = undefined,
-    };
+    while (it.peek().?.id == .LineComment) _ = it.next();
 
-    const tree = try arena.create(ast.Tree);
-    tree.* = ast.Tree{
+    const tree = try arena.create(Tree);
+    tree.* = Tree{
         .source = source,
-        .root_node = root_node,
+        .root_node = undefined,
+        .tokens = token_list,
+        .errors = Tree.ErrorList.init(arena),
         .arena_allocator = tree_arena,
-        .tokens = ast.Tree.TokenList.init(arena),
-        .errors = ast.Tree.ErrorList.init(arena),
     };
+    tree.root_node = try parseRoot(&tree.arena_allocator.allocator, &it, tree);
+    return tree;
+}
 
-    var tokenizer = Tokenizer.init(tree.source);
-    while (true) {
-        const token_ptr = try tree.tokens.addOne();
-        token_ptr.* = tokenizer.next();
-        if (token_ptr.id == Token.Id.Eof) break;
-    }
-    var tok_it = tree.tokens.iterator(0);
+// Root <- skip ContainerMembers eof
+fn parseRoot(arena: *Allocator, it: *TokenIterator, tree: *Tree) Allocator.Error!*Node.Root {
+    const node = try arena.create(Node.Root);
+    node.* = Node.Root{
+        .base = Node{ .id = .Root },
+        .decls = undefined,
+        // TODO: File an issue: Currently it is impossible to have a container-level
+        //       doc comment and NO doc comment on the first decl. For now, simply
+        //       ignore the problem and assume that there will be no container-level
+        //       doc comments.
+        .doc_comments = null,
+        .eof_token = undefined,
+    };
+    node.decls = parseContainerMembers(arena, it, tree, .Keyword_struct) catch |err| {
+        // TODO: Switch on the error type
+        // https://github.com/ziglang/zig/issues/2203
```

Comment:
I went ahead and solved #2203 so that you could remove this TODO comment.

On the other hand, maybe this should be a new proposal because I see here you have this idea that you could have an entire error set as one of the switch prongs. That's a great idea!

Reply:
Done! https://github.com/ziglang/zig/issues/2473